Introduce GenerationConfig

larryliu0820 opened the pull request:
Summary:
Started to implement https://github.com/pytorch/executorch/issues/9341
Started to fix https://github.com/pytorch/executorch/issues/8495

This PR introduces `GenerationConfig` which contains the configs that can be changed across different invocations of `generate()`. 

For example, `temperature` is moved out from the runner constructor for it's not tied to the runner instance but instead should be adjustable every time we call `generate()`.

Similarly we put `echo` and `warming` into the config.

We also allow both `seq_len` and `max_new_tokens` to be passed through the config and we determine the value of `max_new_tokens` based on these 2 config values, pte file metadata as well as the number of prompt tokens.

Differential Revision: D73091676

iseeyuan commented on examples/models/llama/main.cpp:
```diff
@@ -73,13 +73,14 @@ int32_t main(int32_t argc, char** argv) {
 
   if (warmup) {
-    runner.warmup(prompt, seq_len);
+    runner.warmup(prompt, /*max_new_tokens=*/seq_len);
```
Would it be added in the internal runner as well?

larryliu0820 replied:
which internal runner?

iseeyuan commented on extension/llm/runner/irunner.h:
```diff
@@ -6,42 +6,114 @@
+  // Whether to echo the input prompt in the output
+  bool echo = true;
+
+  // Maximum number of total tokens
```
More doc string on this: a limit that users can set. It must not be larger than max_context_len from the model. (something like that, if I understand it correctly).

cccclai commented on examples/mediatek/executor_runner/mtk_llama_runner.h:
```diff
@@ -43,11 +43,9 @@ class MTKLlamaRunner : public executorch::extension::llm::IRunner {
   Error generate(
       const std::string& prompt,
-      int32_t seq_len = 128,
+      executorch::extension::llm::GenerationConfig config,
```
Can we leave mtk code as is? It's currently in unstable stage given the new apk release and upcoming code change.

larryliu0820 committed:
Introduce GenerationConfig (#10228)

Summary:

Started to implement https://github.com/pytorch/executorch/issues/9341
Started to fix https://github.com/pytorch/executorch/issues/8495

This PR introduces `GenerationConfig` which contains the configs that can be changed across different invocations of `generate()`. 

For example, `temperature` is moved out from the runner constructor for it's not tied to the runner instance but instead should be adjustable every time we call `generate()`.

Similarly we put `echo` and `warming` into the config.

We also allow both `seq_len` and `max_new_tokens` to be passed through the config and we determine the value of `max_new_tokens` based on these 2 config values, pte file metadata as well as the number of prompt tokens.

Reviewed By: iseeyuan

Differential Revision: D73091676
